src/main/java/wolox/training/config/CustomAuthenticationProvider.java

```diff
@@ -0,0 +1,44 @@
+package wolox.training.config;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.stereotype.Component;
+import wolox.training.models.Users;
+import wolox.training.repositories.UsersRepository;
+import wolox.training.services.PasswordEncoderService;
+
+import java.util.ArrayList;
+import java.util.Optional;
+
+@Component
+public class CustomAuthenticationProvider implements AuthenticationProvider {
+
+    @Autowired
+    UsersRepository usersRepository;
+
+    @Override
+    public Authentication authenticate(Authentication authentication)
+            throws AuthenticationException {
+        String name = authentication.getName();
+        String password = authentication.getCredentials().toString();
+
+        Optional<Users> user = usersRepository.findByUsername(name);
+
+        boolean matchPassword = PasswordEncoderService.passwordMatch(password, user.get().getPassword());
+
+        if (user.isPresent() && matchPassword) {
+            return new UsernamePasswordAuthenticationToken(
+                    name, password, new ArrayList<>());
+        } else {
```

Comment:
Why dont throw a custom exception here so we can tell what was wrong here??

Reply:
I don't know why or how, but it already do it. Before when I have the matches issue, the response was very clear about what was failing. But if its better practice I can do the custom exception.

Reply:
Is ok, the default exception here can work, you can always create an exception or exception handler.